File: Kernel Simulator/Network/RemoteDebug/RemoteDebugHelpSystem.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO

Public Module RemoteDebugHelpSystem

    'This dictionary is the definitions for commands.
    Public RDebugDefinitions As Dictionary(Of String, String)
    Public RDebugModDefs As New Dictionary(Of String, String)

    ''' <summary>
    ''' Updates the help definition so it reflects the available commands
    ''' </summary>
    Public Sub InitRDebugHelp()
        RDebugDefinitions = New Dictionary(Of String, String) From {{"trace", DoTranslation("Shows last stack trace on exception")},
                                                                    {"username", DoTranslation("Shows current username in the session")},
                                                                    {"register", DoTranslation("Sets device username")},
                                                                    {"exit", DoTranslation("Disconnects you from the debugger")},
                                                                    {"help", DoTranslation("Shows help screen")}}
    End Sub

    ''' <summary>
    ''' Shows the help entry for command.
    ''' </summary>
    ''' <param name="command">Specified command</param>
    Public Sub RDebugShowHelp(command As String, DeviceStream As StreamWriter)

        If command = "" Then
            DeviceStream.WriteLine(DoTranslation("General commands:"))
            For Each cmd As String In RDebugDefinitions.Keys
                DeviceStream.WriteLine("- {0}: {1}", cmd, RDebugDefinitions(cmd))
            Next
            DeviceStream.WriteLine(vbNewLine + DoTranslation("Mod commands:"), True, ColTypes.Neutral)
            If RDebugModDefs.Count = 0 Then DeviceStream.WriteLine(DoTranslation("No mod commands."))
            For Each cmd As String In RDebugModDefs.Keys
                DeviceStream.WriteLine("- {0}: {1}", cmd, RDebugModDefs(cmd))
            Next
            DeviceStream.WriteLine(vbNewLine + DoTranslation("Alias commands:"), True, ColTypes.Neutral)
            If RemoteDebugAliases.Count = 0 Then DeviceStream.WriteLine(DoTranslation("No alias commands."))
            For Each cmd As String In RemoteDebugAliases.Keys
                DeviceStream.WriteLine("- {0}: {1}", cmd, RDebugDefinitions(RemoteDebugAliases(cmd)))
            Next
        ElseIf command = "help" Then
            DeviceStream.WriteLine(DoTranslation("Usage:") + " /help [command]")
        ElseIf command = "trace" Then
            DeviceStream.WriteLine(DoTranslation("Usage:") + " /trace <tracenumber>")
        ElseIf command = "username" Then
            DeviceStream.WriteLine(DoTranslation("Usage:") + " /username")
        ElseIf command = "register" Then
            DeviceStream.WriteLine(DoTranslation("Usage:") + " /register <username>")
        ElseIf command = "exit" Then
            DeviceStream.WriteLine(DoTranslation("Usage:") + " /exit")
        Else
            DeviceStream.WriteLine(DoTranslation("No help for command ""{0}""."), command)
        End If

    End Sub

End Module

File: public/Nitrocid/Kernel/Debugging/RemoteDebug/Command/RemoteDebugHelpSystem.cs

// Nitrocid KS  Copyright (C) 2018-2023  Aptivi
// 
// This file is part of Nitrocid KS
// 
// Nitrocid KS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Nitrocid KS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Linq;
using KS.Languages;
using KS.Shell.ShellBase.Help;

namespace KS.Kernel.Debugging.RemoteDebug.Command
{
    /// <summary>
    /// Help system for shells module
    /// </summary>
    public static class RemoteDebugHelpSystem
    {

        /// <summary>
        /// Shows the list of commands under the current shell type
        /// </summary>
        /// <param name="device">Device to contact</param>
        public static void ShowHelp(RemoteDebugDevice device) =>
            ShowHelp("", device);

        /// <summary>
        /// Shows the help of a command, or command list under the current shell type if nothing is specified
        /// </summary>
        /// <param name="command">A specified command</param>
        /// <param name="device">Device to contact</param>
        public static void ShowHelp(string command, RemoteDebugDevice device)
        {
            // Determine command type
            var CommandList = RemoteDebugCommandExecutor.RemoteDebugCommands
                              .OrderBy((CommandValuePair) => CommandValuePair.Key)
                              .ToDictionary((CommandValuePair) => CommandValuePair.Key, (CommandValuePair) => CommandValuePair.Value);

            // Check to see if command exists
            if (!string.IsNullOrWhiteSpace(command) & CommandList.ContainsKey(command))
            {
                // Found!
                var FinalCommandList = CommandList;
                string FinalCommand = command;
                string RenderedCommand = $"/{command}";
                string HelpDefinition = FinalCommandList[FinalCommand].GetTranslatedHelpEntry();
                int UsageLength = Translate.DoTranslation("Usage:").Length;
                var HelpUsages = Array.Empty<string>();

                // Populate help usages
                if (FinalCommandList[FinalCommand].CommandArgumentInfo is not null)
                    HelpUsages = FinalCommandList[FinalCommand].CommandArgumentInfo.HelpUsages;

                // Print usage information
                foreach (string HelpUsage in HelpUsages)
                    DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, Translate.DoTranslation("Usage:") + $" {FinalCommand} {HelpUsage}", true, device);

                // Write the description now
                if (string.IsNullOrEmpty(HelpDefinition))
                    HelpDefinition = Translate.DoTranslation("Command defined by ") + command;
                DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, Translate.DoTranslation("Description:") + $" {HelpDefinition}", true, device);

                // Extra help action for some commands
                FinalCommandList[FinalCommand].CommandBase?.HelpHelper();
            }
            else if (string.IsNullOrWhiteSpace(command))
            {
                // List the available commands
                if (!HelpPrintTools.SimHelp)
                {
                    // The built-in commands
                    DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, Translate.DoTranslation("General commands:") + (HelpPrintTools.ShowCommandsCount & HelpPrintTools.ShowShellCommandsCount ? " [{0}]" : ""), true, device, CommandList.Count);

                    // Check the command list count and print not implemented. This is an extremely rare situation.
                    if (CommandList.Count == 0)
                        DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, "- " + Translate.DoTranslation("Shell commands not implemented!!!"), true, device);
                    foreach (string cmd in CommandList.Keys)
                        DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, "- {0}: {1}", true, device, cmd, CommandList[cmd].GetTranslatedHelpEntry());
                }
                else
                {
                    // The built-in commands
                    foreach (string cmd in CommandList.Keys)
                        DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, "{0}, ", true, device, cmd);
                }
            }
            else
            {
                DebugWriter.WriteDebugDeviceOnly(DebugLevel.I, Translate.DoTranslation("No help for command \"{0}\"."), true, device, command);
            }
        }

    }
}
